lib/tbot/ca_rotation.go

```diff
@@ -0,0 +1,149 @@
+//   certificates issued by the old CA, and stop trusting the new CA.
+// - Update Servers -> Standby: So we can stop trusting the old CA.
+
+func (b *Bot) caRotationLoop(ctx context.Context) error {
+	// TODO: Throw error if more than X consecutive errors in time window.
+	for ctx.Err() == nil {
+		err := b.watchCARotations(ctx)
+		if err != nil {
+			b.log.WithError(err).Warnf("Error ocurred whilst watching CA rotations, retrying...")
+		}
+	}
+
+	return nil
+}
+
+func (b *Bot) watchCARotations(ctx context.Context) error {
+	clusterName := b.ident().ClusterName
+	b.log.Debugf("Attempting to establish watch for CA events")
+	watcher, err := b.client().NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{{
+			Kind: types.KindCertAuthority,
+			Filter: types.CertAuthorityFilter{
+				types.HostCA:     clusterName,
+				types.UserCA:     clusterName,
+				types.DatabaseCA: clusterName,
+			}.IntoMap(),
+		}},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer watcher.Close()
+
+	for {
+		select {
+		case event := <-watcher.Events():
+			// OpInit is a special case omitted by the Watcher when the
+			// connection succeeds.
+			if event.Type == types.OpInit {
+				b.log.Infof("Started watching for CA rotations")
+				continue
+			}
+
+			ignoreReason := filterCAEvent(b.log, event, clusterName)
+			if ignoreReason != "" {
+				b.log.Debugf("Ignoring CA event: %s", ignoreReason)
+				continue
+			}
+
+			// We need to debounce here, as multiple events will be received if
+			// the user is rotating multiple CAs at once.
+			b.log.Infof("CA Rotation step detected; attempting to force renewal.")
+			select {
+			case b.reloadChan <- struct{}{}:
+			default:
+				// TODO: Come up with a significantly less janky debounce method
+				b.log.Debugf("Renewal already in progress; ignoring event.")
+			}
+		case <-watcher.Done():
+			if err := watcher.Error(); err != nil {
+				return trace.Wrap(err)
+			}
+			return nil
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sliceContains(slice []string, search string) bool {
+	for _, v := range slice {
+		if v == search {
+			return true
+		}
+	}
+	return false
+}
+
+// filterCAEvent returns a reason why an event should be ignored or an empty
+// string is a renewal is needed.
+func filterCAEvent(log logrus.FieldLogger, event types.Event, clusterName string) string {
+	if event.Type != types.OpPut {
+		return "type not PUT"
+	}
+	ca, ok := event.Resource.(types.CertAuthority)
+	if !ok {
+		return fmt.Sprintf("event resource was not CertAuthority (%T)", event.Resource)
+	}
+	log.Debugf("Filtering CA: %+v %s %s", ca, ca.GetKind(), ca.GetSubKind())
+
+	// We want to update for all phases but init and update_servers
+	phase := ca.GetRotation().Phase
+	if sliceContains([]string{"init", "update_servers"}, phase) {
+		return fmt.Sprintf("skipping due to phase: %s", phase)
+	}
+
+	// Skip anything not from our cluster
+	if ca.GetClusterName() != clusterName {
```

Comment:
Checking the cluster name and the CA type here feels redundant as the Watcher should only send us the events we ask for, can I safely remove these checks ?? Or is there known cases where the Watcher does not filter as expected ?

Reply:
@rosstimothy can you answer this one?